.github/workflows/main.yml

```diff
@@ -1,79 +1,233 @@
+
+  test_backends:
+    runs-on: ${{ matrix.os }}
+    needs:
+      - gen_conda_env_file
+    defaults:
+      run:
+        shell: bash -l {0}
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          - windows-latest
+        python-version:
+          - "3.7"
+          - "3.9"
+        pyarrow:
+          - "0.12.1"
+          - "5"
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+
+      - name: download env file
+        uses: actions/download-artifact@v2
+        with:
+          name: conda-env-file
+          path: tmp
+
+      - uses: conda-incubator/setup-miniconda@v2
+        with:
+          mamba-version: "*"
+          miniforge-variant: Mambaforge
+          miniforge-version: latest
+          channel-priority: strict
+          activate-environment: ibis
+          python-version: ${{ matrix.python-version }}
+          environment-file: tmp/environment.yaml
+          condarc-file: ci/condarc
+
+      - name: install pyarrow
+        run: conda install -n ibis pyarrow=${{ matrix.pyarrow }}
+
+      - name: install ibis
+        run: pip install .
+
+      - name: download backend data
+        run: python ci/datamgr.py download
+
+      - name: set backends
+        id: set_backends
+        run: echo '::set-output name=backends::["csv", "dask", "hdf5", "pandas", "parquet", "sqlite"]'
+
+      - name: install backend data
+        run: |
+          set -euo pipefail
+
+          for backend in ${{ join(fromJSON(steps.set_backends.outputs.backends), ' ') }}; do
+            python ci/datamgr.py "$backend"
+          done
+
+      - name: run tests
+        env:
+          PYTEST_BACKENDS: ${{ join(fromJSON(steps.set_backends.outputs.backends), ' ') }}
+        run: |
+          pytest \
+            ibis/backends/{${{ join(fromJSON(steps.set_backends.outputs.backends), ',') }}} \
+            ibis/backends/tests \
+            -m "${{ join(fromJSON(steps.set_backends.outputs.backends), ' or ') }}" \
+            --numprocesses auto \
+            --quiet \
+            -ra \
+            --junitxml=junit.xml \
+            --cov=ibis \
+            --cov-report=xml:coverage.xml | tee pytest.log
+
+      - name: pytest errors
+        run: ./ci/pytest_errors.sh
+        if: failure()
+
+  test_linux_only_backends:
     runs-on: ubuntu-latest
-    env:
-      BACKENDS: "postgres mysql sqlite"
+    needs:
+      - gen_conda_env_file
+    defaults:
+      run:
+        shell: bash -l {0}
     strategy:
       fail-fast: false
       matrix:
-        python_version: ["3.7", "3.9"]
+        python-version:
+          - "3.7"
+          - "3.9"
+        pyspark:
+          - "2.4.3"
+          - "3"
+        pyarrow:
+          - "0.12.1"
+          - "5"
+        exclude:
+          - python-version: "3.7"
+            pyspark: "3"
+            pyarrow: "0.12.1"
+          - python-version: "3.9"
+            pyspark: "3"
+            pyarrow: "0.12.1"
```

Comment:
I personally find this a step backewards to what we've got now. If these versions can be obtained from the pyproject.toml or poetry files, my preference would be to keep what we've got now (files in `ci/deps/...`).